backend/src/schema/resolvers/embeds/scraper.js

```diff
@@ -0,0 +1,102 @@
+import Metascraper from 'metascraper'
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+
+import { ApolloError } from 'apollo-server'
+import isEmpty from 'lodash/isEmpty'
+import isArray from 'lodash/isArray'
+import mergeWith from 'lodash/mergeWith'
+
+const error = require('debug')('embed:error')
+
+const metascraper = Metascraper([
+  require('metascraper-author')(),
+  require('metascraper-date')(),
+  require('metascraper-description')(),
+  require('metascraper-image')(),
+  require('metascraper-lang')(),
+  require('metascraper-lang-detector')(),
+  require('metascraper-logo')(),
+  // require('metascraper-clearbit-logo')(),
+  require('metascraper-publisher')(),
+  require('metascraper-title')(),
+  require('metascraper-url')(),
+  require('metascraper-audio')(),
+  require('metascraper-soundcloud')(),
+  require('metascraper-video')(),
+  require('metascraper-youtube')(),
+
+  // require('./rules/metascraper-embed')()
+])
+
+let oEmbedProvidersFile = fs.readFileSync(path.join(__dirname, './providers.json'), 'utf8')
+
+// some providers allow a format parameter
+// we need JSON
+oEmbedProvidersFile = oEmbedProvidersFile.replace('{format}', 'json')
+
+const oEmbedProviders = JSON.parse(oEmbedProvidersFile)
+
+const fetchEmbed = async url => {
+  const provider = oEmbedProviders.find(provider => {
+    return provider.provider_url.includes(url.hostname)
+  })
+  if (!provider) return {}
+  const {
+    endpoints: [endpoint],
+  } = provider
+  const endpointUrl = new URL(endpoint.url)
+  endpointUrl.searchParams.append('url', url.href)
+  endpointUrl.searchParams.append('format', 'json')
+  let json
+  try {
+    const response = await fetch(endpointUrl)
+    json = await response.json()
+  } catch (err) {
+    error(`Error fetching embed data: ${err.message}`)
+    return {}
+  }
+
+  return {
+    type: json.type,
+    html: json.html,
+    author: json.author_name,
+    date: json.upload_date,
+    sources: ['oembed'],
+  }
+}
+
+const fetchResource = async url => {
+  const response = await fetch(url)
+  const html = await response.text()
+  const resource = await metascraper({ html, url: url.href })
+  return {
+    sources: ['resource'],
+    ...resource,
+  }
+}
+
+export default async function scrape(url) {
+  url = new URL(url)
+  if (url.hostname === 'youtu.be') {
+    // replace youtu.be to get proper results
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:30:37Z" title="Monday, July 22nd 2019, 9:30:37 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,102 @@
+import Metascraper from 'metascraper'
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+
+import { ApolloError } from 'apollo-server'
+import isEmpty from 'lodash/isEmpty'
+import isArray from 'lodash/isArray'
+import mergeWith from 'lodash/mergeWith'
+
+const error = require('debug')('embed:error')
+
+const metascraper = Metascraper([
+  require('metascraper-author')(),
+  require('metascraper-date')(),
+  require('metascraper-description')(),
+  require('metascraper-image')(),
+  require('metascraper-lang')(),
+  require('metascraper-lang-detector')(),
+  require('metascraper-logo')(),
+  // require('metascraper-clearbit-logo')(),
+  require('metascraper-publisher')(),
+  require('metascraper-title')(),
+  require('metascraper-url')(),
+  require('metascraper-audio')(),
+  require('metascraper-soundcloud')(),
+  require('metascraper-video')(),
+  require('metascraper-youtube')(),
+
+  // require('./rules/metascraper-embed')()
+])
+
+let oEmbedProvidersFile = fs.readFileSync(path.join(__dirname, './providers.json'), 'utf8')
+
+// some providers allow a format parameter
+// we need JSON
+oEmbedProvidersFile = oEmbedProvidersFile.replace('{format}', 'json')
+
+const oEmbedProviders = JSON.parse(oEmbedProvidersFile)
+
+const fetchEmbed = async url => {
+  const provider = oEmbedProviders.find(provider => {
+    return provider.provider_url.includes(url.hostname)
+  })
+  if (!provider) return {}
+  const {
+    endpoints: [endpoint],
+  } = provider
+  const endpointUrl = new URL(endpoint.url)
+  endpointUrl.searchParams.append('url', url.href)
+  endpointUrl.searchParams.append('format', 'json')
+  let json
+  try {
+    const response = await fetch(endpointUrl)
+    json = await response.json()
+  } catch (err) {
+    error(`Error fetching embed data: ${err.message}`)
+    return {}
+  }
+
+  return {
+    type: json.type,
+    html: json.html,
+    author: json.author_name,
+    date: json.upload_date,
+    sources: ['oembed'],
+  }
+}
+
+const fetchResource = async url => {
+  const response = await fetch(url)
+  const html = await response.text()
+  const resource = await metascraper({ html, url: url.href })
+  return {
+    sources: ['resource'],
+    ...resource,
+  }
+}
+
+export default async function scrape(url) {
+  url = new URL(url)
+  if (url.hostname === 'youtu.be') {
+    // replace youtu.be to get proper results
```


thank you for your strategic use of comments @roschaefer